File: docker/scripts/converter.go
package main

import (
    "bufio"
    "compress/gzip"
    "encoding/csv"
    "encoding/json"
    "errors"
    "flag"
    "fmt"
    "io"
    "log"
    "net/http"
    "os"
    "path/filepath"
    "runtime"
    "strconv"
    "strings"
    "sync"
    "time"
)

// parser.add_argument("file", nargs='+', help="input file name")
// parser.add_argument("--outdir", default=None)
// parser.add_argument("--gzip", default=False, action="store_true")
// parser.add_argument("--outfile-prefix", default=None)
// args = parser.parse_args()

var (
    pOutDir = flag.String("outdir", "", "")
    pGzipOut = flag.Bool("gzip", true, "")
    pYear = flag.Int("year", 0, "")
    pSource = flag.String("source", "https://data.gharchive.org", "")
    pMonth = flag.Int("month", 0, "")
    pDay = flag.Int("day", 0, "")
)

type Event struct {
    ID string
    Type string
    Actor struct {
        ID int
        Login string
    }
    Repo struct {
        ID int
        Name string
    }
    CreatedAt string `json:"created_at"`
    Payload json.RawMessage
}
func (e *Event) Fields()[]string{
    return []string {
        e.ID,
        strconv.Itoa(e.Actor.ID),
        e.Actor.Login,
        strconv.Itoa(e.Repo.ID),
        e.Repo.Name,
        e.CreatedAt,
    }
}

type PushEventPayload struct {
    Head string
    Before string
    Size int
    DistinctSize int `json:"distinct_size"`
}

func openFile(fname string)(io.ReadCloser, error){
    l := strings.ToLower(fname)
    var reader io.ReadCloser
    var err error
    if strings.HasPrefix(l, "http://") || strings.HasPrefix(l, "https://") {
        resp, err := http.Get(fname)
        if err != nil {
            return nil, err
        }
        if resp.StatusCode != http.StatusOK {
            return nil, errors.New("Bad return code: " + resp.Status)
        }
        reader = resp.Body
    } else {
        reader, err = os.Open(fname)
        if err != nil {
            return nil, err
        }
    }
    
    if strings.HasSuffix(l, ".gz") {
        reader, err = gzip.NewReader(reader)
        if err != nil {
            return nil, err
        }
    }
    
    return reader, nil
}

func writeFields(c *csv.Writer, fields []string)error{
    for fieldIndex, field := range fields{
        for _, sub := range []string{"\\", "\"", ",", ";", "'", "\t", "\x00", "\r", "\n"}{
            field = strings.Replace(field, sub, " ", -1)
        }
        fields[fieldIndex] = field
    }
    return c.Write(fields)
}

func processLine(line []byte, files *OutFiles)error{
    var event Event
    err := json.Unmarshal(line, &event)
    if err != nil {
        return err
    }
    
    switch event.Type {
    case "PushEvent":
        var payload PushEventPayload
        err := json.Unmarshal([]byte(event.Payload), &payload)
        if err != nil {
            return err
        }
        fields := append(event.Fields(), []string{
            payload.Head, payload.Before, strconv.Itoa(payload.Size), strconv.Itoa(payload.DistinctSize),
        }...)
        return writeFields(files.GetWriter(event.Type), fields)
    }
    return nil
}

func processReader(r io.ReadCloser, files *OutFiles) error {
    scanner := bufio.NewScanner(r)
    lineNumber := 0
    for scanner.Scan(){
        lineNumber++
        err := processLine(scanner.Bytes(), files)
        if err != nil {
            return fmt.Errorf("can't process line line '%v': %v", lineNumber, err)
        }
    }
    return nil
}

func dateExist(year, month, day int)bool{
    t := time.Date(year, time.Month(month), day, 0, 0, 0, 0, time.UTC)
    return t.Day() == day && int(t.Month()) == month && t.Year() == year
}

type Stream struct {
    Name string
    InFiles []string
    OutFiles *OutFiles
}

func(s *Stream)String()string{
    return s.Name
}

type OutFiles struct {
    prefix string
    csv map[string]*csv.Writer
    writers []io.Closer
    tmpFiles []string
}

func(o *OutFiles) GetWriter(t string)*csv.Writer{
    if _, ok := o.csv[t]; !ok {
        outFileName := o.prefix + "_" + strings.ToLower(t) + ".csv"
        if *pGzipOut {
            outFileName += ".gz"
        }
        outFileName += ".tmp"
        f, err := os.Create(outFileName)
        if err != nil {
            log.Fatalf("Can't create outfile '%v': %v", outFileName, err)
        }
        o.tmpFiles = append(o.tmpFiles, outFileName)
        o.writers = append(o.writers, f)
        var writer io.WriteCloser = f
        if *pGzipOut {
            writer = gzip.NewWriter(writer)
            o.writers = append(o.writers, writer)
        }
        csvWriter := csv.NewWriter(writer)
        csvWriter.UseCRLF = true
        o.csv[t] = csvWriter
    }
    return o.csv[t]
}

func (o *OutFiles)Close(ok bool){
    for key, c :=range o.csv{
        c.Flush()
        err := c.Error()
        if err != nil {
            log.Println("Error flush csv '%v': %v",key, err)
        }
    }
    
    for i := len(o.writers)-1; i >= 0; i--{
        err := o.writers[i].Close()
        if err != nil {
            log.Println("Error out close: %v", err)
        }
    }
    
    if ok {
        for _, tmpfName := range o.tmpFiles {
            fName := strings.TrimSuffix(tmpfName, ".tmp")
            err := os.Rename(tmpfName, fName)
            if err != nil {
                log.Println("Error rename '%v' -> '%v': %v", tmpfName, fName, err)
            }
        }
    }
}

func isHttp(s string )bool{
    return strings.HasPrefix(s, "http://") || strings.HasPrefix(s, "https://")
}

func streamGenerator() []*Stream {
    
    var res []*Stream
    var minMonth, maxMonth= 1, 12

    minYear, maxYear := 2015, time.Now().Year()
    if *pYear != 0 {
        minYear, maxYear = *pYear, *pYear
    }
    
    if *pMonth != 0 {
        minMonth, maxMonth = *pMonth, *pMonth
    }
    
    var minDay, maxDay= 1, 31
    if *pDay != 0 {
        minDay, maxDay = *pDay, *pDay
    }
    for year := minYear; year <=maxYear; year++ {
        for month := minMonth; month <= maxMonth; month++ {
            for day := minDay; day <= maxDay; day++ {
                if !dateExist(year, month, day) {
                    break
                }
                
                outFilePrefix := fmt.Sprintf("%4d-%02d-%02d", year, month, day)
                
                outFiles := OutFiles{
                    prefix:  filepath.Join(*pOutDir, outFilePrefix),
                    csv:     make(map[string]*csv.Writer),
                    writers: nil,
                }
                var stream = &Stream{
                    Name:     outFilePrefix,
                    OutFiles: &outFiles,
                    InFiles:  make([]string, 0, 24),
                }
                for hour := 0; hour <= 23; hour++ {
                    fName := fmt.Sprintf("%4d-%02d-%02d-%d.json.gz", year, month, day, hour)
                    if isHttp(*pSource) {
                        fName = *pSource + "/" + fName
                    } else {
                        fName = filepath.Join(*pSource, fName)
                    }
                    filePath := fName
                    stream.InFiles = append(stream.InFiles, filePath)
                }
                res = append(res, stream)
            }
        }
    }
    return res
}

func processStream(stream *Stream) {
    processOk := false
    log.Printf("Start stream: %v\n", stream.Name)
    defer func() {
        stream.OutFiles.Close(processOk)
    }()
    
    for _, inFileName := range stream.InFiles{
        inFile, err := openFile(inFileName)
        if err != nil {
            log.Printf("Can't open read file '%v': %v\n", inFile, err)
            return
        }
        err = processReader(inFile, stream.OutFiles)
        if err != nil {
            log.Printf("Error process file '%v' %v\n", inFileName, err)
            return
        }
    }
    processOk = true
    log.Printf("Finish: %v\n", stream.Name)
}

func work(streams []*Stream){
    workerCount := runtime.NumCPU()
    
    var wg sync.WaitGroup
    wg.Add(workerCount)
    
    tasks := make(chan *Stream)
    go func() {
        for _, stream := range streams{
            tasks <- stream
        }
        close(tasks)
    }()
    
    for i := 0; i < workerCount; i++ {
        go func() {
            defer wg.Done()
            
            for stream := range tasks {
                processStream(stream)
            }
        }()
    }
    wg.Wait()
}

func main(){
    flag.Parse()
    
    streams := streamGenerator()
    
    work(streams)
}

File: docker/scripts/converter.py
import argparse
import json
import gzip
import typing
import csv
import os.path
import urllib.request


def val(d: dict, *args: str):
    v = d
    for arg in args:
        if v.get(arg) is None:
            return ""
        v = v[arg]
    return v


def process_raw(line: bytes, get_file: typing.Callable):
    raw = json.loads(line)

    def common_raw() -> list:
        out = [
            raw["id"],
            raw["actor"]["id"],
            raw["actor"]["login"],
            raw["repo"]["id"],
            raw["repo"]["name"],
            raw["created_at"]
        ]
        return out

    t = raw["type"]

    def write_raw(out_raw, event=None):
        if event is None:
            event = t

        f = get_file(event)
        for (index, field_val) in enumerate(out_raw):
            has_replaced = False
            if type(field_val) == str:
                for spec in ("\\", "\"", ",", ";", "'", "\t", "\x00", "\r", "\n"):
                    if spec in field_val:
                        field_val = field_val.replace(spec, " ")
                        has_replaced = True

            if type(field_val) == bool:
                field_val = 1 if field_val else 0
                has_replaced = True

            if has_replaced:
                out_raw[index] = field_val

        f.writerow(out_raw)

    out_raw = common_raw()
    payload = raw["payload"]
    if t == "PushEvent":
        out_raw = common_raw()
        out_raw.extend([payload["head"], payload["before"], payload["size"], payload["distinct_size"]])
        write_raw(out_raw)
        for commit in payload["commits"]:
            out_raw = common_raw()
            out_raw.extend([commit["author"]["name"], commit["author"]["email"], commit["sha"]])
            write_raw(out_raw, "push_commit")
    # if t == "CreateEvent":
    #     out_raw.extend([payload["ref_type"], payload["ref"], payload["master_branch"], payload["description"], payload["pusher_type"]])
    #     write_raw(out_raw)
    if t == "WatchEvent":
        # out_raw.extend([payload["action"]])
        write_raw(out_raw)
    if t == "IssueCommentEvent":
        issue = payload["issue"]
        user = issue["user"]
        comment = payload["comment"]
        out_raw.extend([
            payload["action"],
            issue["id"],
            issue["number"],
            issue["html_url"],
            issue["title"],
            issue["state"],
            issue["created_at"],
            user["login"],
            user["id"],
            comment["id"],
            comment["body"]
        ])
        write_raw(out_raw)
    if t == "PullRequestEvent":
        pr = payload["pull_request"]
        base = pr["base"]
        head = pr["head"]

        head_forks = val(head, "repo", "forks")

        def supress_empty_int(v):
            if v == "":
                return 0
            return v

        out_raw.extend([
            pr["number"],
            payload["action"],
            pr["url"],
            pr["title"],
            pr["state"],
            pr["body"],
            pr["merged"],
            pr["comments"],
            pr["commits"],

            val(base, "repo", "full_name"),
            val(head, "repo", "full_name"),

            val(base, "repo", "owner"),
            val(head, "repo", "owner"),

            val(base, "repo", "license", "key"),
            val(head, "repo", "license", "key"),

            supress_empty_int(val(base, "repo", "forks")),
            supress_empty_int(val(head, "repo", "forks")),
        ])
        write_raw(out_raw)

    if t == "IssuesEvent":
        issue = payload["issue"]
        labels = [label["name"].replace(";", ".,") for label in payload["issue"]["labels"]]
        out_raw.extend([
            payload["action"],
            issue["id"],
            issue["number"],
            issue["html_url"],
            issue["title"],
            issue["state"],
            issue["created_at"],
            issue["updated_at"],
            issue["user"]["login"],
            ";".join(labels),
        ])
        write_raw(out_raw)

    if t == "ForkEvent":
        forkee = payload["forkee"]
        out_raw.extend([
            forkee["name"],
            forkee["owner"]["login"],
        ])
        write_raw(out_raw)

outfiles = {}
def get_file(eventtype: str, outfile_prefix:str, gzipout=True)-> typing.IO:
    key = outfile_prefix + "_" + eventtype
    if key not in outfiles:
        outfile_name = outfile_prefix + "_" + eventtype.lower() + ".csv"
        if gzipout:
            outfile_name += ".gz"
            f = gzip.open(outfile_name, "wt", encoding="utf8", newline="")
        else:
            f = open(outfile_name, "wt", encoding="utf8", newline="")
        f_csv = csv.writer(f, quoting=csv.QUOTE_NONNUMERIC)
        outfiles[key] = f_csv
    return outfiles[key]


def process_file(fname: str, gzipout=True, outdir=None, outfile_prefix=None):

    if fname.lower().startswith("http://") or fname.lower().startswith("https://"):
        finput = urllib.request.urlopen(fname)
    else:
        finput = open(fname, mode="rb")

    if fname.lower().endswith(".gz"):
        finput = gzip.GzipFile(fileobj=finput)

    if outfile_prefix is None:
        outfile_prefix = fname

    if outfile_prefix.endswith(".gz"):
        outfile_prefix = outfile_prefix[:-len(".gz")]
    if outfile_prefix.endswith(".json"):
        outfile_prefix = outfile_prefix[:-len(".json")]
    if outdir is not None:
        outfile_prefix = os.path.join(outdir, os.path.basename(outfile_prefix))

    with finput:
        line_number=0
        while True:
            line = finput.readline()
            if not line:
                return
            line_number += 1
            process_raw(line, lambda eventtype: get_file(eventtype, outfile_prefix, gzipout))


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("file", nargs='+', help="input file name")
    parser.add_argument("--outdir", default=None)
    parser.add_argument("--gzip", default=False, action="store_true")
    parser.add_argument("--outfile-prefix", default=None)
    args = parser.parse_args()
    for file in args.file:
        print("Start process:", file)
        process_file(file, gzipout=args.gzip, outdir=args.outdir, outfile_prefix=args.outfile_prefix)


if __name__ == '__main__':
    main()
